File: java/src/main/java/dev/oneslang/Lexer.java
package dev.oneslang;

import java.util.ArrayList;
import java.util.List;

/**
 * 1S: ERP Free Edition — Lexer (Java)
 * Tokenises UTF-8 source with bilingual Russian/English keywords.
 */
public class Lexer {

    private final String source;
    private final String filename;
    private int pos = 0, line = 1, col = 1;

    public Lexer(String source, String filename) {
        this.source   = source;
        this.filename = filename;
    }

    // ── Public ──────────────────────────────────────────────────────────────

    public List<Token> tokenize() {
        var tokens = new ArrayList<Token>();
        while (pos < source.length()) {
            Token t = nextToken();
            if (t != null) tokens.add(t);
        }
        tokens.add(new Token(TokenType.EOF, null, line, col));
        return tokens;
    }

    // ── Helpers ──────────────────────────────────────────────────────────────

    private char peek() { return pos < source.length() ? source.charAt(pos) : '\0'; }
    private char peek(int off) {
        int i = pos + off;
        return i < source.length() ? source.charAt(i) : '\0';
    }

    private char advance() {
        char c = source.charAt(pos++);
        if (c == '\n') { line++; col = 1; } else { col++; }
        return c;
    }

    private int[] curPos() { return new int[]{line, col}; }

    // ── Scanner ───────────────────────────────────────────────────────────────

    private Token nextToken() {
        // Skip whitespace and newlines
        while (pos < source.length() && " \t\r\n".indexOf(peek()) >= 0) advance();
        if (pos >= source.length()) return null;

        int[] lc = curPos();
        char c = peek();

        // Comment
        if (c == '/' && peek(1) == '/') return readComment(lc);

        // Preprocessor
        if (c == '#') return readPreprocessor(lc);

        // String
        if (c == '"' || c == '\'') return readString(lc, c);

        // Number
        if (Character.isDigit(c)) return readNumber(lc);

        // Identifier / keyword (Latin or Cyrillic)
        if (Character.isLetter(c) || c == '_') return readWord(lc);

        // Operators
        return readOperator(lc);
    }

    private Token readComment(int[] lc) {
        var buf = new StringBuilder();
        while (pos < source.length() && peek() != '\n') buf.append(advance());
        return new Token(TokenType.COMMENT, buf.toString(), lc[0], lc[1]);
    }

    private Token readPreprocessor(int[] lc) {
        var buf = new StringBuilder();
        buf.append(advance()); // '#'
        while (pos < source.length() && (Character.isLetter(peek()) || peek() == '_'))
            buf.append(advance());
        String word = buf.toString().toLowerCase();
        // Argument (e.g. region name)
        var arg = new StringBuilder();
        while (pos < source.length() && peek() != '\n' && peek() != '\r')
            arg.append(advance());
        // Resolve via a minimal inline table (matches Python PP_KEYWORDS)
        TokenType tt = switch (word) {
            case "#если",       "#if"        -> TokenType.PP_IF;
            case "#тогда",      "#then"      -> TokenType.PP_THEN;
            case "#иначе",      "#else"      -> TokenType.PP_ELSE;
            case "#иначеесли",  "#elseif"    -> TokenType.PP_ELSIF;
            case "#конецесли",  "#endif"     -> TokenType.PP_ENDIF;
            case "#область",    "#region"    -> TokenType.PP_REGION;
            case "#конецобласти","#endregion"-> TokenType.PP_ENDREGION;
            case "#использовать","#use"      -> TokenType.PP_USE;
            default                          -> null;
        };
        if (tt == null) return null;
        return new Token(tt, arg.toString().strip(), lc[0], lc[1]);
    }

    private Token readString(int[] lc, char quote) {
        advance(); // opening quote
        var buf = new StringBuilder();
        while (pos < source.length()) {
            char ch = peek();
            if (ch == quote) {
                advance();
                if (peek() == quote) { buf.append(advance()); } // doubled escape
                else break;
            } else if (ch == '\n') {
                advance();
                while (pos < source.length() && (peek() == ' ' || peek() == '\t')) advance();
                if (peek() == '|') { advance(); buf.append('\n'); }
                else throw new RuntimeException("Unterminated string at " + lc[0] + ":" + lc[1]);
            } else {
                buf.append(advance());
            }
        }
        String val = buf.toString();
        if (val.matches("\\d{8}")) return new Token(TokenType.DATE, val, lc[0], lc[1]);
        return new Token(TokenType.STRING, val, lc[0], lc[1]);
    }

    private Token readNumber(int[] lc) {
        var buf = new StringBuilder();
        while (pos < source.length() && Character.isDigit(peek())) buf.append(advance());
        if (pos < source.length() && peek() == '.' && Character.isDigit(peek(1))) {
            buf.append(advance());
            while (pos < source.length() && Character.isDigit(peek())) buf.append(advance());
            return new Token(TokenType.FLOAT, buf.toString(), lc[0], lc[1]);
        }
        return new Token(TokenType.INTEGER, buf.toString(), lc[0], lc[1]);
    }

    private Token readWord(int[] lc) {
        var buf = new StringBuilder();
        while (pos < source.length() &&
               (Character.isLetterOrDigit(peek()) || peek() == '_'))
            buf.append(advance());
        String word = buf.toString();
        TokenType tt = TokenType.lookup(word);
        return new Token(tt, word, lc[0], lc[1]);
    }

    private Token readOperator(int[] lc) {
        char c = advance();
        return switch (c) {
            case '+' -> new Token(TokenType.PLUS,      "+", lc[0], lc[1]);
            case '-' -> new Token(TokenType.MINUS,     "-", lc[0], lc[1]);
            case '*' -> new Token(TokenType.MULTIPLY,  "*", lc[0], lc[1]);
            case '/' -> new Token(TokenType.DIVIDE,    "/", lc[0], lc[1]);
            case '%' -> new Token(TokenType.MODULO,    "%", lc[0], lc[1]);
            case '(' -> new Token(TokenType.LPAREN,    "(", lc[0], lc[1]);
            case ')' -> new Token(TokenType.RPAREN,    ")", lc[0], lc[1]);
            case '[' -> new Token(TokenType.LBRACKET,  "[", lc[0], lc[1]);
            case ']' -> new Token(TokenType.RBRACKET,  "]", lc[0], lc[1]);
            case '{' -> new Token(TokenType.LBRACE,    "{", lc[0], lc[1]);
            case '}' -> new Token(TokenType.RBRACE,    "}", lc[0], lc[1]);
            case ',' -> new Token(TokenType.COMMA,     ",", lc[0], lc[1]);
            case ';' -> new Token(TokenType.SEMICOLON, ";", lc[0], lc[1]);
            case '.' -> new Token(TokenType.DOT,       ".", lc[0], lc[1]);
            case '?' -> new Token(TokenType.QUESTION,  "?", lc[0], lc[1]);
            case ':' -> new Token(TokenType.COLON,     ":", lc[0], lc[1]);
            case '=' -> new Token(TokenType.EQ,        "=", lc[0], lc[1]);
            case '<' -> {
                if (peek() == '>') { advance(); yield new Token(TokenType.NEQ, "<>", lc[0], lc[1]); }
                if (peek() == '=') { advance(); yield new Token(TokenType.LTE, "<=", lc[0], lc[1]); }
                yield new Token(TokenType.LT, "<", lc[0], lc[1]);
            }
            case '>' -> {
                if (peek() == '=') { advance(); yield new Token(TokenType.GTE, ">=", lc[0], lc[1]); }
                yield new Token(TokenType.GT, ">", lc[0], lc[1]);
            }
            default -> throw new RuntimeException(
                String.format("Unknown char '%c' at %d:%d", c, lc[0], lc[1]));
        };
    }
}

File: src/lexer/lexer.py
"""
1S: ERP Free Edition — Lexer
Handles bilingual (Russian / English) source files, UTF-8.
"""
from __future__ import annotations
import re
from typing import Iterator
from .tokens import Token, TokenType, KEYWORDS, PP_KEYWORDS


class LexerError(Exception):
    def __init__(self, msg: str, line: int, col: int):
        super().__init__(f"[Lexer] {msg} at {line}:{col}")
        self.line = line
        self.col = col


class Lexer:
    def __init__(self, source: str, filename: str = "<unknown>"):
        self.source = source.lstrip("﻿")  # strip UTF-8 BOM if present
        self.filename = filename
        self.pos = 0
        self.line = 1
        self.col = 1
        self._tokens: list[Token] = []

    # ── Public API ──────────────────────────────────────────────────────────

    def tokenize(self) -> list[Token]:
        tokens = list(self._scan())
        tokens.append(Token(TokenType.EOF, None, self.line, self.col))
        return tokens

    # ── Internal scanner ────────────────────────────────────────────────────

    def _scan(self) -> Iterator[Token]:
        while self.pos < len(self.source):
            ch = self._peek()

            # Skip whitespace (not newlines — newlines are significant for
            # statement boundaries in some 1C dialects, but we treat ; as
            # the primary separator and skip newlines too)
            if ch in " \t\r":
                self._advance()
                continue

            if ch == "\n":
                self._advance()
                continue

            # Comment: //
            if ch == "/" and self._peek(1) == "/":
                yield self._read_comment()
                continue

            # Preprocessor directive  #...
            if ch == "#":
                tok = self._read_preprocessor()
                if tok:
                    yield tok
                continue

            # String literal  "..."  or  '...'  (1C uses double-quotes)
            if ch in ('"', "'"):
                yield self._read_string(ch)
                continue

            # Date literal  '19991231'  (single-quoted 8-digit date)
            # Handled inside _read_string — detected by content

            # Number
            if ch.isdigit():
                yield self._read_number()
                continue

            # Identifier / keyword (Latin or Cyrillic)
            if ch.isalpha() or ch == "_":
                yield self._read_word()
                continue

            # Operators and punctuation
            tok = self._read_operator()
            if tok:
                yield tok
                continue

            raise LexerError(f"Unexpected character {ch!r}", self.line, self.col)

    # ── Character helpers ────────────────────────────────────────────────────

    def _peek(self, offset: int = 0) -> str:
        idx = self.pos + offset
        return self.source[idx] if idx < len(self.source) else "\0"

    def _advance(self) -> str:
        ch = self.source[self.pos]
        self.pos += 1
        if ch == "\n":
            self.line += 1
            self.col = 1
        else:
            self.col += 1
        return ch

    def _current_pos(self) -> tuple[int, int]:
        return self.line, self.col

    # ── Token readers ────────────────────────────────────────────────────────

    def _read_comment(self) -> Token:
        line, col = self._current_pos()
        buf = []
        while self.pos < len(self.source) and self._peek() != "\n":
            buf.append(self._advance())
        return Token(TokenType.COMMENT, "".join(buf), line, col)

    def _read_preprocessor(self) -> Token | None:
        line, col = self._current_pos()
        buf = [self._advance()]  # consume '#'
        while self.pos < len(self.source) and (self._peek().isalpha() or self._peek() in "_"):
            buf.append(self._advance())
        text = "".join(buf).lower()
        from .tokens import PP_KEYWORDS
        ttype = PP_KEYWORDS.get(text)
        if ttype is None:
            # Unknown preprocessor directive — skip rest of line
            while self.pos < len(self.source) and self._peek() != "\n":
                self._advance()
            return None
        # For #Region / #Use collect the argument
        arg = ""
        if ttype in (TokenType.PP_REGION, TokenType.PP_USE):
            while self.pos < len(self.source) and self._peek() not in ("\n", "\r"):
                arg += self._advance()
            arg = arg.strip()
        return Token(ttype, arg or None, line, col)

    def _read_string(self, quote: str) -> Token:
        line, col = self._current_pos()
        self._advance()  # opening quote
        buf = []
        while self.pos < len(self.source):
            ch = self._peek()
            if ch == quote:
                self._advance()
                # 1C: doubled quote = escape  "" → "
                if self._peek() == quote:
                    buf.append(self._advance())
                else:
                    break
            elif ch == "\n":
                # Multi-line strings: 1C uses | as line continuation prefix
                self._advance()
                # skip leading whitespace then expect |
                while self._peek() in (" ", "\t"):
                    self._advance()
                if self._peek() == "|":
                    self._advance()
                    buf.append("\n")
                else:
                    raise LexerError("Unterminated string literal", line, col)
            else:
                buf.append(self._advance())
        value = "".join(buf)
        # Detect 1C date literal: 8 digits exactly
        if re.fullmatch(r"\d{8}", value):
            return Token(TokenType.DATE, value, line, col)
        return Token(TokenType.STRING, value, line, col)

    def _read_number(self) -> Token:
        line, col = self._current_pos()
        buf = []
        while self._peek().isdigit():
            buf.append(self._advance())
        if self._peek() == "." and self._peek(1).isdigit():
            buf.append(self._advance())  # '.'
            while self._peek().isdigit():
                buf.append(self._advance())
            return Token(TokenType.FLOAT, float("".join(buf)), line, col)
        return Token(TokenType.INTEGER, int("".join(buf)), line, col)

    def _read_word(self) -> Token:
        line, col = self._current_pos()
        buf = []
        while self.pos < len(self.source) and (self._peek().isalnum() or self._peek() == "_"):
            buf.append(self._advance())
        word = "".join(buf)
        lower = word.lower()
        ttype = KEYWORDS.get(lower, TokenType.IDENTIFIER)
        return Token(ttype, word, line, col)

    def _read_operator(self) -> Token | None:
        line, col = self._current_pos()
        ch = self._advance()

        single = {
            "+": TokenType.PLUS,
            "-": TokenType.MINUS,
            "*": TokenType.MULTIPLY,
            "/": TokenType.DIVIDE,
            "%": TokenType.MODULO,
            "(": TokenType.LPAREN,
            ")": TokenType.RPAREN,
            "[": TokenType.LBRACKET,
            "]": TokenType.RBRACKET,
            "{": TokenType.LBRACE,
            "}": TokenType.RBRACE,
            ",": TokenType.COMMA,
            ";": TokenType.SEMICOLON,
            ".": TokenType.DOT,
            "?": TokenType.QUESTION,
            ":": TokenType.COLON,
            "~": TokenType.LABEL,
        }

        if ch == "<":
            if self._peek() == ">":
                self._advance()
                return Token(TokenType.NEQ, "<>", line, col)
            if self._peek() == "=":
                self._advance()
                return Token(TokenType.LTE, "<=", line, col)
            return Token(TokenType.LT, "<", line, col)

        if ch == ">":
            if self._peek() == "=":
                self._advance()
                return Token(TokenType.GTE, ">=", line, col)
            return Token(TokenType.GT, ">", line, col)

        if ch == "=":
            return Token(TokenType.EQ, "=", line, col)

        if ch in single:
            return Token(single[ch], ch, line, col)

        raise LexerError(f"Unknown operator character {ch!r}", line, col)
